src/ros_filter.cpp

```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
 
     // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
     // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
-    // the queue.
+    // the queue. If the history is insufficiently short, we just take the oldest state we have.
+    FilterStatePtr lastHistoryState;
     while (!filterStateHistory_.empty() && filterStateHistory_.back()->lastMeasurementTime_ > time)
     {
+      lastHistoryState = filterStateHistory_.back();
       filterStateHistory_.pop_back();
     }
 
+    // If the state history is not empty at this point, it means that our history was large enough, and we
+    // should revert to the state at the back of the history deque.
+    bool retVal = false;
+    if (!filterStateHistory_.empty())
+    {
+      retVal = true;
+      lastHistoryState = filterStateHistory_.back();
+    }
+    else
+    {
+      RF_DEBUG("Insufficient history to revert to time " << time << "\n");
```

Comment:
Most of the logic for this history management (including that) came from a contributor PR, actually.

I can't check the `front`, necessarily, because then I'd have to pop to be able to look at the next value, and I don't want to do that. This is a `deque`, though, which supports random access, so we can walk to the point in question and then do an `erase` on the rest. However, I wouldn't want to start at the `0` index/front because we're more likely to be forced to search through more results to get to the time in question than if we start from the back of the deque and work towards the front..

Reply:
However, it looks like `pop_back` is a constant-time operation, so given that this is working, I'm inclined to leave it alone, unless you feel strongly about it.

Reply:
I don't feel strongly about it, no. And to be clear, this suggestion is only as an alternative to running through every item in the queue to decide if we have a long enough history -- I definitely agree that working backwards from the most recent measurement is more sensible for finding the last relevant measurement.

Reply:
Oooh, I see, you mean just doing an immediate check on the length of the queue, gotcha.

Reply:
Yeah. You have the option of just looking at how much time the queue spans directly. I don't know how much of a benefit that is, but I guess the more inputs you have (and the higher frequency they are) the more pops you may have to do unnecessarily.

Reply:
But that would only be true if the queue is too short, right? Once it fills up even a little bit, we'll almost never get a measurement that exceeds its bounds. For example, we have it set to 0.4 seconds, and we violate it maybe once in the entire lifetime of the node (right at the beginning of execution).

Also, a `clear` on the deque is [O(n)](http://en.cppreference.com/w/cpp/container/deque/clear), as is the act of popping every value in it. I'm sure popping is more like 2n, but I'm personally not fussed, so long as they're both linear.

Reply:
Yeah, definitely sounds like it wouldn't be worth it.